File: houseRobber3.rb
def rob(root)
  rob_helper(root).max
end

def rob_helper(root)
  return [0, 0] if root.nil?

  left_taken, left_not_taken = rob_helper(root.left)
  right_taken, right_not_taken = rob_helper(root.right)

  [root.val + left_not_taken + right_not_taken, [left_taken, left_not_taken].max + [right_taken, right_not_taken].max]
end

File: old_problems/house_robber3/houseRobber3.js
var rob = function(root) {
  let money = helper(root);

  return Math.max(money[0], money[1]);
};

var helper = function(root) {
  if (!root) { return [0,0]; }

  let left = helper(root.left);
  let right = helper(root.right);
  let res = [0,0];
  res[0] = Math.max(left[0], left[1]) + Math.max(right[0], right[1]);
  res[1] = root.val + left[0] + right[0];

  return res;
};
